_posts/grant_guo/2017-07-07-spark.md

```diff
@@ -0,0 +1,74 @@
+---
+layout: post
+title:  "A performance-tuning practice: from 20 hours to 8 minutes"
+breadcrumb: true
+author: grant_guo
+date: 2017-07-06
+categories: grant_guo
+tags:
+    - Spark SQL
+teaser:
+    info: An introduction of how to tune DCC-Release ID job and reduce the execution time greatly
+    image: spark-logo-trademark.png
+header:
+    version: small
+    title: Software Engineering Blog
+    image: header-logo-crop.png
+    icon: icon-blog
+---
+
+## Background
+
+In our product environment, the DCC-Release ID job takes 20+ hours to finish. The bottleneck is the fact that for every data point ID job needs to talk to DCC-ID service through a HTTP request to either obtain or create an ID, which means every time the ID job runs there are millions of ID lookup requests, even if ID service runs multiple instances, and the vanish cache runs on a dedicated server with 128G RAM.
+
+## Solution
+
+The ID service already has a "****/export" endpoint for each ID type(mutation, sample, donor and specimen) respectively, so the original idea is to cache the exported data at worker nodes locally for lookups and only make the HTTP request when a new ID is required to be generated. 
+
+The first trial is to use Spark Broadcast(https://spark.apache.org/docs/latest/programming-guide.html#broadcast-variables), which is used to give every node a copy of a large input dataset in an efficient manner. 
```

Comment:
replace "which is used to give every node"  by "which gives every node", this also prevent the duplication of "use" in the sentence.